docs/dev/topology-over-raft.md

```diff
@@ -294,6 +300,8 @@ stateDiagram-v2
     use_new --> cleanup
     cleanup --> end_migration
     end_migration --> [*]
+    [*] --> repair
+    repair --> [*]
```

Comment:
I'm having second thoughts about this. 

I think we should have `end_repair` stage, which will call the barrier before clearing the transition. `session` would be cleared when setting stage to `end_repair`. The reason is that this guarantees that if tablet is not in transition there is no repair process running in the system (barrier waits for it). Without this, it is possible that repair is left behind on error. 

Other topology operations, like tablet split finalization may then run concurrently with those abandoned processes, and cause some issues.

It doesn't seem like much effort to introduce `end_repair` stage, and it would make it easier for us to reason about correctness of the system because we will have this strong invariant.

Reply:
How about this:

1)tablet_transition_stage::repair:
set_stage(end_repair) 
del_session()

2)tablet_transition_stage::end_repair
do_barrier()
del_transition()
del_tablet_task_info()
set_repair_time()


```
                case locator::tablet_transition_stage::repair: {
                    if (action_failed(tablet_state.repair)) {
                        if (do_barrier()) {
                            _tablets.erase(gid);
                            updates.emplace_back(get_mutation_builder()
                                    .del_transition(last_token)
                                    .build());
                        }
                        break;
                    }
                    if (advance_in_background(gid, tablet_state.repair, "repair", [&] () -> future<> {
                        auto& tinfo = tmap.get_tablet_info(gid.tablet);
                        auto sched_time = tinfo.tablet_task_info.sched_time;
                        auto primary = tmap.get_primary_replica(gid.tablet);
                        auto dst = primary.host;
                        auto tablet = gid;
                        rtlogger.info("Initiating tablet repair host={} tablet={}", dst, gid);
                        co_await ser::storage_service_rpc_verbs::send_tablet_repair(&_messaging,
                                netw::msg_addr(id2ip(dst)), _as, raft::server_id(dst.uuid()), gid);
                        auto duration = std::chrono::duration<float>(db_clock::now() - sched_time);
                        rtlogger.info("Finished tablet repair host={} tablet={} duration={}", dst, tablet, duration);
                    })) {
                            rtlogger.debug("Will set tablet {} stage to {}", gid, locator::tablet_transition_stage::end_repair);
                            updates.emplace_back(get_mutation_builder()
                                .set_stage(last_token, locator::tablet_transition_stage::end_repair)
                                .del_session(last_token)
                                .build());

                    }
                }
                    break;
                case locator::tablet_transition_stage::end_repair: {
                    if (do_barrier()) {
                        auto& tinfo = tmap.get_tablet_info(gid.tablet);
                        auto time = tinfo.tablet_task_info.sched_time;
                        _tablets.erase(gid);
                        updates.emplace_back(get_mutation_builder()
                            .del_transition(last_token)
                            .del_tablet_task_info(last_token)
                            .set_repair_time(last_token, time)
                            .build());
                    }
                }
                    break;
            }
```

Reply:
When action failed, here:

```
            case locator::tablet_transition_stage::repair: {
                    if (action_failed(tablet_state.repair)) {
                        if (do_barrier()) {
                            _tablets.erase(gid);
                            updates.emplace_back(get_mutation_builder()
                                    .del_transition(last_token)
                                    .build());
```
you need to jump to `end_repair` because we want the barrier after clearing session and before deleting the transition.

Then `repair_time` should be set in the repair stage, only in the success path. `end_repair` is both on the success and failure path.

Reply:
OK. I changed it in the new version.